### PR TITLE
SW-350 one time show a guide to the user if laser head s model is detected

### PR DESCRIPTION
**Pull request by khaledsherkawi**

### CODE REVIEW

File: octoprint_mrbeam/static/js/laserhead_changed.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    function LaserheadChangedViewModel(params) {
+        let self = this;
+        window.mrbeam.viewModels["LaserheadChangedViewModel"] = self;
+
+        self.settings = params[0];
+        self.loginState = params[1];
+
+        self.onUserLoggedIn = function () {
+            if (
+                self.loginState.currentUser?.()?.active
```
Comment (amrsoll) [inline octoprint_mrbeam/static/js/laserhead_changed.js:11]: looks interesting, could `currentUser` not exist? Do you need the extra `?.` ?

File: octoprint_mrbeam/static/js/laserhead_changed.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            ) {
+                let laserheadChanged = self.settings.settings.plugins.mrbeam.laserheadChanged();
+                if (laserheadChanged) {
```
Comment (amrsoll) [inline octoprint_mrbeam/static/js/laserhead_changed.js]: Seems like it can be shortened

File: octoprint_mrbeam/static/js/laserhead_changed.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                                "Unable to save laserhead change detection at the moment.%(br)sCheck connection to Mr Beam and try again."
+                            ),
+                            { br: "<br/>" }
+                        ),
+                        type: "error",
+                        hide: true,
+                    });
+                });
+        }
+    }
```
Comment (amrsoll) [inline octoprint_mrbeam/static/js/laserhead_changed.js]: What is the purpose of this? Why is the front end telling the back-end whether the laserhead got changed?

Comment (amrsoll) [reply octoprint_mrbeam/static/js/laserhead_changed.js]: If the data is static... Couldn't we ignore it all together?

File: octoprint_mrbeam/static/js/laserhead_changed.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
+$(function () {
+    function LaserheadChangedViewModel(params) {
```
Comment (amrsoll) [inline octoprint_mrbeam/static/js/laserhead_changed.js:2]: I think you will need to argument for why there is a separate viewmodel for this feature.

File: octoprint_mrbeam/static/js/laserhead_changed.js (hunk 3 above)
Comment (amrsoll) [reply octoprint_mrbeam/static/js/laserhead_changed.js]: I think I understand what this is about now... Could you rename it to `laserheadChangeAcknowledged()` ?

File: octoprint_mrbeam/__init__.py [hunk 5]
```diff
@@ -1918,6 +1926,8 @@ def on_api_command(self, command, data):
             return NO_CONTENT
         elif command == "focus_reminder":
             return self.focus_reminder(data)
+        elif command == "laserhead_changed":
+            return self.laserhead_changed(data)
```
Comment (amrsoll) [inline octoprint_mrbeam/__init__.py]: ... so, in the backend, you can receive the user acknowledgment, and then decide to 
```py
self._settings.set_boolean(["laserheadChanged"], dict_get(data, "laserhead_change_acknowledged", False)
```
I do not see the need for a separate `laserhead_changed()` function.

**Review by amrsoll — CHANGES_REQUESTED** [review]

File: octoprint_mrbeam/static/js/laserhead_changed.js (hunk 1 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/static/js/laserhead_changed.js:11]: Yes I think that happens in the scenario when the session expires

File: octoprint_mrbeam/static/js/laserhead_changed.js (hunk 2 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/static/js/laserhead_changed.js]: done

File: octoprint_mrbeam/static/js/laserhead_changed.js (hunk 3 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/static/js/laserhead_changed.js]: I was think about Acknowledged as well ^^ 
Done

File: octoprint_mrbeam/static/js/laserhead_changed.js (hunk 4 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/static/js/laserhead_changed.js:2]: Well that's the way we are handling this for now (check focus reminder). But I agree we need to revamp this at some point and have all such windows belong to one View model.

File: octoprint_mrbeam/__init__.py (hunk 5 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/__init__.py]: I'm not sure what you mean here

Comment (amrsoll) [reply octoprint_mrbeam/__init__.py]: basically it was shortening it too

Comment (amrsoll) [reply octoprint_mrbeam/__init__.py]: nvm

File: octoprint_mrbeam/static/js/laserhead_changed.js (hunk 3 above)
Comment (amrsoll) [reply octoprint_mrbeam/static/js/laserhead_changed.js]: Could you also reflect the new name in the json you send please?

Comment (amrsoll) [reply octoprint_mrbeam/static/js/laserhead_changed.js]: Or do not send any "data", just send back empty data with the API call `laserhead_change_acknowledged`

Comment (khaledsherkawi) [reply octoprint_mrbeam/static/js/laserhead_changed.js]: No I think that stands correct. The config parameter reflect an event happening which is laser being changed

Comment (amrsoll) [reply octoprint_mrbeam/static/js/laserhead_changed.js]: And what is that for?
`self.settings.requestData();`

Comment (amrsoll) [reply octoprint_mrbeam/static/js/laserhead_changed.js]: Yes, in the config itself, not in the front-end. The front end only knows that the user acknowledged the change, but it is up to the backend to choose whether to do something with the acknowledgement

Comment (khaledsherkawi) [reply octoprint_mrbeam/static/js/laserhead_changed.js]: self.settings.requestData() has been removed

Comment (khaledsherkawi) [reply octoprint_mrbeam/static/js/laserhead_changed.js]: I agree with you, I just don't see what the need to change the parameter name into acknowledged in the backend

**Review by amrsoll — APPROVED** [review]

**Pull request merged**